File: R/tests/testthat/test_utilities.R
# Copyright (c) Facebook, Inc. and its affiliates.

# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

library(prophet)
context("Prophet utilities tests")

DATA <- read.csv('data.csv')
DATA$ds <- as.Date(DATA$ds)

build_model_with_regressors <- function(data, mcmc.samples = 0) {
  m <- prophet(mcmc.samples = mcmc.samples)
  m <- add_regressor(m, 'binary_feature', prior.scale=0.2)
  m <- add_regressor(m, 'numeric_feature', prior.scale=0.5)
  m <- add_regressor(
    m, 'numeric_feature2', prior.scale=0.5, mode = 'multiplicative')
  m <- add_regressor(m, 'binary_feature2', standardize=TRUE)

  df <- data
  df$binary_feature <- c(rep(0, 255), rep(1, 255))
  df$numeric_feature <- 0:509
  df$numeric_feature2 <- 0:509
  df$binary_feature2 <- c(rep(1, 100), rep(0, 410))
  m <- fit.prophet(m, df)

  return(m)
}

test_that("regressor_coefficients_no_uncertainty", {
  skip_if_not(Sys.getenv('R_ARCH') != '/i386')
  m <- build_model_with_regressors(DATA, mcmc.samples = 0)
  coefs <- regressor_coefficients(m)

  expect_equal(dim(coefs), c(4, 6))
  expect_equal(coefs[, "coef_lower"], coefs[, "coef"])
  expect_equal(coefs[, "coef_upper"], coefs[, "coef"])
})

test_that("regressor_coefficients_with_uncertainty", {
  skip_if_not(Sys.getenv('R_ARCH') != '/i386')
  suppressWarnings(m <- build_model_with_regressors(DATA, mcmc.samples = 100))
  coefs <- regressor_coefficients(m)

  expect_equal(dim(coefs), c(4, 6))
  expect_true(all(coefs[, "coef_lower"] < coefs[, "coef"]))
  expect_true(all(coefs[, "coef_upper"] > coefs[, "coef"]))
})

File: python/prophet/tests/test_utilities.py
# Copyright (c) Facebook, Inc. and its affiliates.

# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import numpy as np

from prophet import Prophet
from prophet.utilities import regressor_coefficients


class TestUtilities:
    def test_regressor_coefficients(self, daily_univariate_ts, backend):
        m = Prophet(stan_backend=backend)
        df = daily_univariate_ts.copy()
        # Set fixed seed for reproducibility
        np.random.seed(123)
        # Generate random values once and assign to both columns
        random_values = np.random.normal(size=df.shape[0])
        df["regr1"] = random_values
        # Reuse same random generator for second column to avoid another call
        df["regr2"] = np.random.normal(size=df.shape[0])
        m.add_regressor("regr1", mode="additive")
        m.add_regressor("regr2", mode="multiplicative")
        m.fit(df)

        # Get coefficients once and use for all assertions
        coefs = regressor_coefficients(m)
        assert coefs.shape == (2, 6)
        # Use numpy array comparison without redundant function calls
        coef_values = coefs["coef"].values
        assert np.array_equal(coefs["coef_lower"].values, coef_values)
        assert np.array_equal(coefs["coef_upper"].values, coef_values)
